src/grade_recognition_systems.asm

```diff
@@ -0,0 +1,383 @@
+    dw $0001 ;00 — 8
+    dw $0002 ;00 — 7
+    dw $0003 ;00 — 6
+    dw $0005 ;00 — 5
+    dw $0007 ;00 — 4
+    dw $0013 ;00 — 3
+    dw $0020 ;00 — 2
+    dw $0030 ;00 — 1
+    dw $0040 ;00 — S1
+    dw $0055 ;00 — S2
+    dw $0075 ;00 — S3
+    dw $0100 ;00 — S4
+    dw $0130 ;00 — S5
+    dw $0165 ;00 — S6
+    dw $0205 ;00 — S7
+    dw $0250 ;00 — S8
+    dw $0300 ;00 — S9
+; Man..., the TGM3 Grade system is really complex...
+sInternalGradeSystem:
+    db 125, 10, 20, 40, 50 ;Decay rate, (Internal grade points awarded for:) Single, Double, Triple, Tetris
+    db 80, 10, 20, 30, 40
+    db 80, 10, 20, 30, 40
+    db 50, 10, 15, 30, 40
+    db 45, 5, 15, 20, 40
+    db 45, 5, 15, 20, 30
+    db 45, 5, 10, 20, 30
+    db 40, 5, 10, 15, 30
+    db 40, 5, 10, 15, 30
+    db 40, 5, 10, 15, 30
+    db 40, 2, 12, 13, 30
+    db 40, 2, 12, 13, 30
+    db 30, 2, 12, 13, 30
+    db 30, 2, 12, 13, 30
+    db 30, 2, 12, 13, 30
+    db 20, 2, 12, 13, 30
+    db 20, 2, 12, 13, 30
+    db 20, 2, 12, 13, 30
+    db 20, 2, 12, 13, 30
+    db 20, 2, 12, 13, 30
+    db 15, 2, 12, 13, 30
+    db 15, 2, 12, 13, 30
+    db 15, 2, 12, 13, 30
+    db 15, 2, 12, 13, 30
+    db 15, 2, 12, 13, 30
+    db 15, 2, 12, 13, 30
+    db 15, 2, 12, 13, 30
+    db 15, 2, 12, 13, 30
+    db 15, 2, 12, 13, 30
+    db 15, 2, 12, 13, 30
+    db 10, 2, 12, 13, 30
+    db 10, 2, 12, 13, 30
+
+sGradeBoosts:
+    ;This should explain itself
+    db 0
+    db 1
+    db 2
+    db 3
+    db 4 
+    db 5 
+    db 5 
+    db 6 
+    db 6 
+    db 7 
+    db 7 
+    db 7 
+    db 8 
+    db 8 
+    db 8 
+    db 9 
+    db 9 
+    db 10 
+    db 11 
+    db 12 
+    db 12 
+    db 12 
+    db 13 
+    db 13 
+    db 14 
+    db 14 
+    db 15
+    db 15 
+    db 16 
+    db 16 
+    db 17
+sComboMultipliers:
+    db 1, 1, 1, 1, 1 ; Combo size, (Multiplier for: ) Single, Double, Triple, Tetris
+    db 2, 1, 1.2, 1.4, 1.5
+    db 3, 1, 1.2, 1.5, 1.8
+    db 4, 1, 1.4, 1.6, 2
+    db 5, 1, 1.4, 1.7, 2.2
+    db 6, 1, 1.4, 1.8, 2.3
+    db 7, 1, 1.4, 1.9, 2.4
+    db 8, 1, 1.5, 2, 2.5
+    db 9, 1, 1.5, 2.1, 2.6
+    db 10, 2, 2.5, 3
+sLevelMultiplier:
+    db 1 ; 000-249
+    db 2 ; 250-499
+    db 3 ; 500-749
+    db 4 ; 750-999
+sBaselineCOOL:
+    db 52 ;070 (value in seconds)
+    db 52 ;170
+    db 49 ;270
+    db 45 ;370
+    db 45 ;470
+    db 42 ;570
+    db 42 ;670
+    db 38 ;770
+    db 38 ;870
+sREGRETConditions:
+    db 1, 30 ;minutes, seconds
+    db 1, 15
+    db 1, 15 
+    db 1, 8 
+    db 1, 0
+    db 1, 0 
+    db 50 ;seconds
+    db 50 
+    db 50 
+    db 50
+
+SECTION "Grading Functions", ROM0
+    ; Wipe the grading variables.
+GradeInit::
+    xor a, a
+    ld [wDecayRate], a
+    ld [wGradePoints], a
+    ld [wInternalGrade], a
+    ld [wDisplayedGrade], a
+    ld a, $7
+    ld [wPalette], a
+    jp UpdateGrade
+    ret
+
+
+    ; Gets the highest grade the player qualifies for.
+UpdateGrade::
+    ; Is the Speed Curve Death or Shirase?
+    ld hl, .GradeJumps

```

Comment:
For future expandability, instead do something like this:

```
    ld hl, .gradejumps
    ld a, [wSpeedCurveState]
    ld b, a
    add a, b
    add a, b
    ld b, 0
    ld c, a
    add hl, bc
    jp hl

.gradejumps
    jp UpdateGradeTGM1 ; DMGT
    jp UpdateGradeTGM1 ; TGM1
    jp UpdateGradeTGM1 ; TGM3
    jp UpdateGradeDEAT ; DEAT
    jp UpdateGradeSHIR ; SHIR
    jp UpdateGradeTGM1 ; CHIL
```

This way, each speed curve is its own entry in the jump table and you don't need to do fragile maths to single out shirase and death modes, and adding extra modes will be easy down the line.

Reply:
true